### ISSUE TITLE
Default restartable linear variant

### ISSUE THREAD

**Issue by mphoward** [glotzerlab/hoomd-blue#486]
## Description

The default `zero` parameter for `hoomd.variant.linear_interp` is `'now'`, which means that the current timestep is considered 0. Since this parameter is pretty analogous to `phase`, I would have expected the default to instead be 0, as it is elsewhere in hoomd. Without specifying `zero=0`, restartable jobs do not work properly, and the resulting output might be confusing to the user. I propose changing the default to `zero=0` rather than `zero='now'` to avoid this confusion.

## Script

```python
kT = hoomd.variant.linear_interp(((0,1),(1000,2)))
# use kT in an integrator...
hoomd.run_upto(1000)
```

## Output

If the job got restarted at step 500, the values of kT would be:
```
# timestep kT
0 1.0
500 1.5
# restart
500 1.0
1000 1.5
```

## Expected output

With `zero=0`, the values of kT should be:
```
# timestep kT
0 1.0
500 1.5
# restart
500 1.5
1000 2.0
```

## Configuration
- **HOOMD-blue** version: 2.6.0 (affects current version)

## Developer

Sure, this is a quick fix if change is wanted by others.

**Comment by joaander** [comment]
Yes, we should make this change. This breaks existing API, so it goes in v3.

We may need to remove the `zero` arg entirely as in the new v3 OO API there is no notion of a global current timestep and `now` becomes ill-defined.

**Comment by mphoward** [comment]
OK, sounds good. I will leave this open then, and you can ping me if you want the change made onto `next`. I honestly don't see much value in the `zero` option generally, since it is straightforward to shift all of the points in the list by an offset if that's the behavior the user wants. I would think having a prescribed time-varying quantity (with a fixed time origin) is much more common.

**Comment by joaander** [comment]
Thanks, Mike. We hope to start making serious progress on the API over the next few weeks. We'll let you know when we have a solid foundation and will most certainly appreciate your help then.

**Comment by joaander** [comment]
Thinking more outside the box: is `variant.linear_interp` sufficiently clean and flexible? Should we keep it? Should we replace it or add a more general interface that the user could customize along with a few simpler built-ins. Something like this:
```
# linear variation from start to end over steps and then hold
variant.Ramp(start=2.0, end=0.5, steps=1e6)
# Cycle from start to end over steps, then from end to start over steps, and repeat
variant.Cycle(start=0.4, end=0.6, steps=1e6)
# user defined
class User(hoomd.variant.Variant):
    def __init__(self, s):
        hoomd.variant.Variant.__init__(self);  # required to initialize C++ parent object
        self.s = s;
   def __call__(self, timestep):
       return math.cos(timestep * self.s)
```

This proposal is modeled after what @b-butler and I designed for the new `Trigger` API used to determine when to trigger `Analyzer` and `Updater` invocations. https://github.com/glotzerlab/hoomd-blue/blob/feature/new-object-API/hoomd/pytest/test_trigger.py

**Comment by mphoward** [comment]
I think this could also be a valid solution. I believe Jim Swan’s group has implemented several variants like this already for their hydrodynamics code, and having the ability to easily implement new variants would be convenient. Do you envision these still being relative time counters or absolute?

**Comment by joaander** [comment]
The timestep will be absolute. This keeps the meaning of `timestep` clear across all places it is used in the API.

With the new API model, there is actually no way to provide the old `zero='now'` logic. A variant can be specified before a simulation context is even created. Variants that wish to provide a phase offset can do so with (timestep - phase) internally. E.g.:
```
variant.Ramp(start=2.0, end=0.5, steps=1e6, phase=100000)
```
This is just a quick example: Some more careful thought is needed to come up with the right terms for these so the meaning is clear without requiring extensive documentation.

**Comment by mphoward** [comment]
Great, then I think your proposal is more self-documenting and robust than the existing variant. Let me know if I can help flesh this out more when you’re ready.

**Comment by joaander** [comment]
Excellent. We would appreciate your help. You can start on this now if you would like. We are currently working on other parts of the API, and Variant could easily be worked on in parallel. Follow the pattern we established with Trigger (Trigger.h, Trigger.cc, triggers.py, pytest/test_trigger.py).

Note: The reason I created a Python `PeriodicTrigger` that inherits from the C++ `_hoomd.PeriodicTrigger` is for sphinx documentation. I'd prefer if we could build the sphinx documentation from the pure python files and not require compiling hoomd's C++ code on readthedocs (there is a dummy `PeriodicTrigger` in `_hoomd.py` for this purpose). We just haven't added docstrings to the classes in `triggers.py` yet.

Also note that I'm building up new unit tests using pytest for the new API. Eventually we will convert and remove the UnitTest based tests, but for now they mostly fail. To run the pytest tests, execute `python -m pytest --pyargs hoomd` in the build directory (later I may add a `make test` target).

**Comment by mphoward** [comment]
I chatted with Jens at AIChE, and I'm going to try to prioritize finishing the new lbvh branch first. Given other constraints, I probably won't be able to work on this for a few weeks, so don't let me hold you up--but happy to help if this is still open then.

**Comment by joaander** [comment]
Implemented in #589
